tests/integration/conftest.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Conftest file for integration tests."""
+
+import functools
+import logging
+import os
+from collections import defaultdict
+from datetime import datetime
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+store = defaultdict(str)
+
+
+def timed_memoizer(func):
+    """Cache the result of a function."""
+
+    @functools.wraps(func)
+    async def wrapper(*args, **kwargs):
+        fname = func.__qualname__
+        logger.info("Started: %s" % fname)
+        start_time = datetime.now()
+        if fname in store.keys():
+            ret = store[fname]
+        else:
+            logger.info("Return for {} not cached".format(fname))
+            ret = await func(*args, **kwargs)
+            store[fname] = ret
+        logger.info("Finished: {} in: {} seconds".format(fname, datetime.now() - start_time))
+        return ret
+
+    return wrapper
+
+
+@pytest.fixture(scope="module")
+@timed_memoizer
+async def charm(ops_test: OpsTest) -> str:
+    """Charm used for integration testing."""
+    if charm_file := os.environ.get("CHARM_PATH"):
+        return str(charm_file)
+
+    charm = await ops_test.build_charm(".")
```

Comment:
I wonder if we should/could:
- Require to always receive `CHARM_PATH` from the outside
- Avoid building charms as part of itests

Doing this would prevent us from using:
- in-charm "tester-charms", but personally I'd rather not use them
- multiple charms from source in one itest, but this is a charm test, not a solution test

Reply:
The problem with that is, if you make some changes to the charm and forget to `pack`, it's very hard to realize that's the problem. Passing it explicitly reduces that risk (if you open a new shell, that env is gone).

It's also slightly worse UX for external people I guess.

I don't strongly have opinions on this, but I'd rather leave this discussion outside of this PR, as I believe it's mostly a CI discussion.